File: server/routes/friends.js
/**
 * Friend System Routes
 * =====================
 * Social features: friend requests, friend lists, online presence,
 * game invites, and blocking.
 *
 * Routes:
 *   GET    /friends              — List friends (accepted)
 *   GET    /friends/requests     — Pending friend requests
 *   GET    /friends/online       — Online friends
 *   POST   /friends/request      — Send friend request
 *   POST   /friends/:id/accept   — Accept a friend request
 *   POST   /friends/:id/decline  — Decline a friend request
 *   POST   /friends/:id/remove   — Remove a friend
 *   POST   /friends/:id/block    — Block a player
 *   POST   /friends/:id/unblock  — Unblock a player
 *   GET    /friends/blocked       — List blocked players
 *   POST   /friends/:id/invite   — Invite friend to a game room
 *   GET    /friends/search       — Search for players by name
 */

const express = require('express');
const db = require('../models/db');

const router = express.Router();

// =========================================
// List Friends
// =========================================

router.get('/', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;

        const result = await db.query(`
            SELECT p.id, p.display_name, p.avatar_character,
                   pp.status as presence_status, pp.current_game_id,
                   f.created_at as friends_since
            FROM friendships f
            JOIN players p ON (
                CASE WHEN f.player_id = $1 THEN p.id = f.friend_id
                     ELSE p.id = f.player_id END
            )
            LEFT JOIN player_presence pp ON pp.player_id = p.id
            WHERE f.tenant_id = $2
                AND (f.player_id = $1 OR f.friend_id = $1)
                AND f.status = 'accepted'
            ORDER BY pp.status = 'online' DESC, p.display_name ASC
        `, [playerId, tenantId]);

        res.json({ friends: result.rows });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Pending Friend Requests
// =========================================

router.get('/requests', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;

        const [incoming, outgoing] = await Promise.all([
            db.query(`
                SELECT f.id, f.player_id as from_id, p.display_name as from_name,
                       p.avatar_character, f.created_at
                FROM friendships f
                JOIN players p ON p.id = f.player_id AND p.tenant_id = f.tenant_id
                WHERE f.friend_id = $1 AND f.tenant_id = $2 AND f.status = 'pending'
                ORDER BY f.created_at DESC
            `, [playerId, tenantId]),
            db.query(`
                SELECT f.id, f.friend_id as to_id, p.display_name as to_name,
                       p.avatar_character, f.created_at
                FROM friendships f
                JOIN players p ON p.id = f.friend_id AND p.tenant_id = f.tenant_id
                WHERE f.player_id = $1 AND f.tenant_id = $2 AND f.status = 'pending'
                ORDER BY f.created_at DESC
            `, [playerId, tenantId])
        ]);

        res.json({
            incoming: incoming.rows,
            outgoing: outgoing.rows
        });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Online Friends
// =========================================

router.get('/online', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;

        const result = await db.query(`
            SELECT p.id, p.display_name, p.avatar_character,
                   pp.status as presence_status, pp.current_game_id, pp.current_room_id
            FROM friendships f
            JOIN players p ON (
                CASE WHEN f.player_id = $1 THEN p.id = f.friend_id
                     ELSE p.id = f.player_id END
            )
            JOIN player_presence pp ON pp.player_id = p.id AND pp.status != 'offline'
            WHERE f.tenant_id = $2
                AND (f.player_id = $1 OR f.friend_id = $1)
                AND f.status = 'accepted'
            ORDER BY pp.status ASC, p.display_name ASC
        `, [playerId, tenantId]);

        res.json({ online: result.rows });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Send Friend Request
// =========================================

router.post('/request', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;
        const { friendId } = req.body;

        if (!friendId) {
            return res.status(400).json({ error: 'friendId is required' });
        }

        if (friendId === playerId) {
            return res.status(400).json({ error: 'Cannot friend yourself' });
        }

        // Check target player exists
        const target = await db.query(
            'SELECT id FROM players WHERE id = $1 AND tenant_id = $2',
            [friendId, tenantId]
        );
        if (target.rows.length === 0) {
            return res.status(404).json({ error: 'Player not found' });
        }

        // Check for existing relationship
        const existing = await db.query(`
            SELECT id, status FROM friendships
            WHERE tenant_id = $1
                AND ((player_id = $2 AND friend_id = $3) OR (player_id = $3 AND friend_id = $2))
        `, [tenantId, playerId, friendId]);

        if (existing.rows.length > 0) {
            const rel = existing.rows[0];
            if (rel.status === 'accepted') {
                return res.status(409).json({ error: 'Already friends' });
            }
            if (rel.status === 'pending') {
                return res.status(409).json({ error: 'Friend request already pending' });
            }
            if (rel.status === 'blocked') {
                return res.status(403).json({ error: 'Cannot send friend request' });
            }
        }

        const result = await db.query(`
            INSERT INTO friendships (tenant_id, player_id, friend_id, status)
            VALUES ($1, $2, $3, 'pending')
            RETURNING id, status, created_at
        `, [tenantId, playerId, friendId]);

        res.status(201).json({
            message: 'Friend request sent',
            request: result.rows[0]
        });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Accept Friend Request
// =========================================

router.post('/:id/accept', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;
        const friendshipId = req.params.id;

        const result = await db.query(`
            UPDATE friendships SET status = 'accepted', updated_at = NOW()
            WHERE id = $1 AND friend_id = $2 AND tenant_id = $3 AND status = 'pending'
            RETURNING id, player_id, friend_id, status
        `, [friendshipId, playerId, tenantId]);

        if (result.rows.length === 0) {
            return res.status(404).json({ error: 'Friend request not found' });
        }

        res.json({ message: 'Friend request accepted', friendship: result.rows[0] });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Decline Friend Request
// =========================================

router.post('/:id/decline', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;
        const friendshipId = req.params.id;

        const result = await db.query(
            'DELETE FROM friendships WHERE id = $1 AND friend_id = $2 AND tenant_id = $3 AND status = $4 RETURNING id',
            [friendshipId, playerId, tenantId, 'pending']
        );

        if (result.rows.length === 0) {
            return res.status(404).json({ error: 'Friend request not found' });
        }

        res.json({ message: 'Friend request declined' });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Remove Friend
// =========================================

router.post('/:id/remove', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;
        const friendId = req.params.id;

        const result = await db.query(`
            DELETE FROM friendships
            WHERE tenant_id = $1 AND status = 'accepted'
                AND ((player_id = $2 AND friend_id = $3) OR (player_id = $3 AND friend_id = $2))
            RETURNING id
        `, [tenantId, playerId, friendId]);

        if (result.rows.length === 0) {
            return res.status(404).json({ error: 'Friendship not found' });
        }

        res.json({ message: 'Friend removed' });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Block / Unblock
// =========================================

router.post('/:id/block', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;
        const targetId = req.params.id;

        // Remove any existing friendship first
        await db.query(`
            DELETE FROM friendships
            WHERE tenant_id = $1
                AND ((player_id = $2 AND friend_id = $3) OR (player_id = $3 AND friend_id = $2))
        `, [tenantId, playerId, targetId]);

        // Create block entry
        await db.query(`
            INSERT INTO friendships (tenant_id, player_id, friend_id, status)
            VALUES ($1, $2, $3, 'blocked')
            ON CONFLICT (tenant_id, player_id, friend_id) DO UPDATE SET status = 'blocked', updated_at = NOW()
        `, [tenantId, playerId, targetId]);

        res.json({ message: 'Player blocked' });
    } catch (err) {
        next(err);
    }
});

router.post('/:id/unblock', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;
        const targetId = req.params.id;

        await db.query(`
            DELETE FROM friendships
            WHERE tenant_id = $1 AND player_id = $2 AND friend_id = $3 AND status = 'blocked'
        `, [tenantId, playerId, targetId]);

        res.json({ message: 'Player unblocked' });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Blocked List
// =========================================

router.get('/blocked', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;

        const result = await db.query(`
            SELECT p.id, p.display_name, p.avatar_character, f.created_at as blocked_at
            FROM friendships f
            JOIN players p ON p.id = f.friend_id AND p.tenant_id = f.tenant_id
            WHERE f.player_id = $1 AND f.tenant_id = $2 AND f.status = 'blocked'
            ORDER BY f.created_at DESC
        `, [playerId, tenantId]);

        res.json({ blocked: result.rows });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Invite Friend to Game
// =========================================

router.post('/:id/invite', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;
        const friendId = req.params.id;
        const { roomId, gameId } = req.body;

        if (!roomId || !gameId) {
            return res.status(400).json({ error: 'roomId and gameId required' });
        }

        // Verify friendship
        const friendship = await db.query(`
            SELECT id FROM friendships
            WHERE tenant_id = $1 AND status = 'accepted'
                AND ((player_id = $2 AND friend_id = $3) OR (player_id = $3 AND friend_id = $2))
        `, [tenantId, playerId, friendId]);

        if (friendship.rows.length === 0) {
            return res.status(403).json({ error: 'Not friends with this player' });
        }

        // The invite will be delivered via WebSocket; just return success
        // (the caller handles WS delivery)
        res.json({
            message: 'Invite sent',
            invite: {
                from: playerId,
                to: friendId,
                roomId,
                gameId,
                createdAt: new Date().toISOString()
            }
        });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Search Players
// =========================================

router.get('/search', async (req, res, next) => {
    try {
        const tenantId = req.player.tenantId;
        const playerId = req.player.id;
        const q = req.query.q || '';
        const limit = Math.min(parseInt(req.query.limit, 10) || 20, 50);

        if (q.length < 2) {
            return res.status(400).json({ error: 'Search query must be at least 2 characters' });
        }

        const result = await db.query(`
            SELECT p.id, p.display_name, p.avatar_character,
                   pp.status as presence_status,
                   (SELECT status FROM friendships f
                    WHERE f.tenant_id = $1
                        AND ((f.player_id = $2 AND f.friend_id = p.id) OR (f.player_id = p.id AND f.friend_id = $2))
                    LIMIT 1) as friendship_status
            FROM players p
            LEFT JOIN player_presence pp ON pp.player_id = p.id
            WHERE p.tenant_id = $1 AND p.id != $2
                AND p.display_name ILIKE $3
            ORDER BY p.display_name ASC
            LIMIT $4
        `, [tenantId, playerId, `%${q}%`, limit]);

        res.json({ players: result.rows });
    } catch (err) {
        next(err);
    }
});

module.exports = router;

File: server-rs/src/routes/friends.rs
use axum::{
    extract::{Path, Query, State},
    Json,
};
use serde::Deserialize;
use serde_json::{json, Value};
use uuid::Uuid;

use crate::error::{AppError, AppResult};
use crate::middleware::auth::AuthPlayer;
use crate::middleware::tenant::TenantId;
use crate::AppState;

#[derive(Deserialize)]
pub struct SearchQuery {
    pub q: Option<String>,
}

pub async fn list_friends(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
) -> AppResult<Json<Value>> {
    let rows: Vec<(Uuid, String, String)> = sqlx::query_as(
        r#"SELECT p.id, p.display_name, p.avatar_character
        FROM friendships f
        JOIN players p ON (
            (f.player_id = $1 AND p.id = f.friend_id) OR
            (f.friend_id = $1 AND p.id = f.player_id)
        ) AND p.tenant_id = $2
        WHERE f.tenant_id = $2 AND f.status = 'accepted'
            AND (f.player_id = $1 OR f.friend_id = $1)"#,
    )
    .bind(player.id)
    .bind(&tenant.0 .0)
    .fetch_all(&state.db)
    .await?;

    let friends: Vec<Value> = rows.iter().map(|(id, name, avatar)| {
        json!({"playerId": id, "displayName": name, "avatarCharacter": avatar})
    }).collect();

    Ok(Json(json!({ "friends": friends })))
}

pub async fn friend_requests(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
) -> AppResult<Json<Value>> {
    let incoming: Vec<(Uuid, String, chrono::DateTime<chrono::Utc>)> = sqlx::query_as(
        r#"SELECT p.id, p.display_name, f.created_at
        FROM friendships f JOIN players p ON p.id = f.player_id AND p.tenant_id = f.tenant_id
        WHERE f.friend_id = $1 AND f.tenant_id = $2 AND f.status = 'pending'"#,
    )
    .bind(player.id)
    .bind(&tenant.0 .0)
    .fetch_all(&state.db)
    .await?;

    let outgoing: Vec<(Uuid, String, chrono::DateTime<chrono::Utc>)> = sqlx::query_as(
        r#"SELECT p.id, p.display_name, f.created_at
        FROM friendships f JOIN players p ON p.id = f.friend_id AND p.tenant_id = f.tenant_id
        WHERE f.player_id = $1 AND f.tenant_id = $2 AND f.status = 'pending'"#,
    )
    .bind(player.id)
    .bind(&tenant.0 .0)
    .fetch_all(&state.db)
    .await?;

    Ok(Json(json!({
        "incoming": incoming.iter().map(|(id, name, created)| json!({"playerId": id, "displayName": name, "createdAt": created})).collect::<Vec<_>>(),
        "outgoing": outgoing.iter().map(|(id, name, created)| json!({"playerId": id, "displayName": name, "createdAt": created})).collect::<Vec<_>>(),
    })))
}

pub async fn online_friends(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
) -> AppResult<Json<Value>> {
    let rows: Vec<(Uuid, String, String, Option<String>, Option<String>)> = sqlx::query_as(
        r#"SELECT p.id, p.display_name, COALESCE(pp.status, 'offline'), pp.current_game_id, pp.current_room_id
        FROM friendships f
        JOIN players p ON (
            (f.player_id = $1 AND p.id = f.friend_id) OR
            (f.friend_id = $1 AND p.id = f.player_id)
        ) AND p.tenant_id = $2
        LEFT JOIN player_presence pp ON pp.player_id = p.id AND pp.tenant_id = $2
        WHERE f.tenant_id = $2 AND f.status = 'accepted' AND (f.player_id = $1 OR f.friend_id = $1)
            AND pp.status IS NOT NULL AND pp.status != 'offline'"#,
    )
    .bind(player.id)
    .bind(&tenant.0 .0)
    .fetch_all(&state.db)
    .await?;

    let friends: Vec<Value> = rows.iter().map(|(id, name, status, gid, rid)| {
        json!({"playerId": id, "displayName": name, "status": status, "currentGameId": gid, "currentRoomId": rid})
    }).collect();

    Ok(Json(json!({ "friends": friends })))
}

pub async fn send_request(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let target_id = body["playerId"].as_str()
        .ok_or_else(|| AppError::BadRequest("playerId required".into()))?;
    let target = Uuid::parse_str(target_id)
        .map_err(|_| AppError::BadRequest("Invalid player ID".into()))?;
    let tid = &tenant.0 .0;

    if target == player.id {
        return Err(AppError::BadRequest("Cannot friend yourself".into()));
    }

    // Check existing
    let existing: bool = sqlx::query_scalar(
        r#"SELECT EXISTS(SELECT 1 FROM friendships WHERE tenant_id = $1
            AND ((player_id = $2 AND friend_id = $3) OR (player_id = $3 AND friend_id = $2)))"#,
    )
    .bind(tid).bind(player.id).bind(target)
    .fetch_one(&state.db).await?;

    if existing {
        return Err(AppError::Conflict("Friendship already exists".into()));
    }

    sqlx::query(
        "INSERT INTO friendships (tenant_id, player_id, friend_id, status, created_at) VALUES ($1, $2, $3, 'pending', NOW())",
    )
    .bind(tid).bind(player.id).bind(target)
    .execute(&state.db).await?;

    Ok(Json(json!({"success": true})))
}

pub async fn accept_request(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Path(id): Path<String>,
) -> AppResult<Json<Value>> {
    let from = Uuid::parse_str(&id).map_err(|_| AppError::BadRequest("Invalid ID".into()))?;
    sqlx::query(
        "UPDATE friendships SET status = 'accepted' WHERE tenant_id = $1 AND player_id = $2 AND friend_id = $3 AND status = 'pending'",
    )
    .bind(&tenant.0 .0).bind(from).bind(player.id)
    .execute(&state.db).await?;
    Ok(Json(json!({"success": true})))
}

pub async fn decline_request(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Path(id): Path<String>,
) -> AppResult<Json<Value>> {
    let from = Uuid::parse_str(&id).map_err(|_| AppError::BadRequest("Invalid ID".into()))?;
    sqlx::query(
        "DELETE FROM friendships WHERE tenant_id = $1 AND player_id = $2 AND friend_id = $3 AND status = 'pending'",
    )
    .bind(&tenant.0 .0).bind(from).bind(player.id)
    .execute(&state.db).await?;
    Ok(Json(json!({"success": true})))
}

pub async fn remove_friend(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Path(id): Path<String>,
) -> AppResult<Json<Value>> {
    let friend = Uuid::parse_str(&id).map_err(|_| AppError::BadRequest("Invalid ID".into()))?;
    sqlx::query(
        r#"DELETE FROM friendships WHERE tenant_id = $1 AND status = 'accepted'
            AND ((player_id = $2 AND friend_id = $3) OR (player_id = $3 AND friend_id = $2))"#,
    )
    .bind(&tenant.0 .0).bind(player.id).bind(friend)
    .execute(&state.db).await?;
    Ok(Json(json!({"success": true})))
}

pub async fn block_player(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Path(id): Path<String>,
) -> AppResult<Json<Value>> {
    let target = Uuid::parse_str(&id).map_err(|_| AppError::BadRequest("Invalid ID".into()))?;
    let tid = &tenant.0 .0;
    // Remove existing friendship
    sqlx::query(
        r#"DELETE FROM friendships WHERE tenant_id = $1
            AND ((player_id = $2 AND friend_id = $3) OR (player_id = $3 AND friend_id = $2))"#,
    )
    .bind(tid).bind(player.id).bind(target)
    .execute(&state.db).await?;
    // Insert block
    sqlx::query(
        "INSERT INTO friendships (tenant_id, player_id, friend_id, status, created_at) VALUES ($1, $2, $3, 'blocked', NOW()) ON CONFLICT DO NOTHING",
    )
    .bind(tid).bind(player.id).bind(target)
    .execute(&state.db).await?;
    Ok(Json(json!({"success": true})))
}

pub async fn unblock_player(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Path(id): Path<String>,
) -> AppResult<Json<Value>> {
    let target = Uuid::parse_str(&id).map_err(|_| AppError::BadRequest("Invalid ID".into()))?;
    sqlx::query(
        "DELETE FROM friendships WHERE tenant_id = $1 AND player_id = $2 AND friend_id = $3 AND status = 'blocked'",
    )
    .bind(&tenant.0 .0).bind(player.id).bind(target)
    .execute(&state.db).await?;
    Ok(Json(json!({"success": true})))
}

pub async fn blocked_list(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
) -> AppResult<Json<Value>> {
    let rows: Vec<(Uuid, String)> = sqlx::query_as(
        r#"SELECT p.id, p.display_name FROM friendships f
        JOIN players p ON p.id = f.friend_id AND p.tenant_id = f.tenant_id
        WHERE f.tenant_id = $1 AND f.player_id = $2 AND f.status = 'blocked'"#,
    )
    .bind(&tenant.0 .0).bind(player.id)
    .fetch_all(&state.db).await?;

    let blocked: Vec<Value> = rows.iter().map(|(id, name)| json!({"playerId": id, "displayName": name})).collect();
    Ok(Json(json!({ "blocked": blocked })))
}

pub async fn invite_to_game(
    State(_state): State<AppState>,
    _player: axum::Extension<AuthPlayer>,
    Path(id): Path<String>,
) -> AppResult<Json<Value>> {
    // In production this would send via WebSocket
    Ok(Json(json!({"success": true, "message": "Invite sent", "targetId": id})))
}

pub async fn search_players(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Query(q): Query<SearchQuery>,
) -> AppResult<Json<Value>> {
    let search = format!("%{}%", q.q.as_deref().unwrap_or(""));

    let rows: Vec<(Uuid, String, String)> = sqlx::query_as(
        "SELECT id, display_name, avatar_character FROM players WHERE tenant_id = $1 AND id != $2 AND display_name ILIKE $3 LIMIT 20",
    )
    .bind(&tenant.0 .0).bind(player.id).bind(&search)
    .fetch_all(&state.db).await?;

    let results: Vec<Value> = rows.iter().map(|(id, name, avatar)| {
        json!({"playerId": id, "displayName": name, "avatarCharacter": avatar})
    }).collect();

    Ok(Json(json!({ "players": results })))
}
